main/src/ca/uwaterloo/flix/language/phase/optimizer/Inliner.scala

```diff
@@ -527,10 +527,42 @@ object Inliner {
     * Returns `true` if
     *   - the local context shows that we are not currently inlining and
     *   - `defn` is not recursive and
+    *   - is either a higher-order function with a known lambda as argument or
     *   - is a direct call to another function.
+    *
+    * It is the responsibility of the caller to visit `exps` first.
     */
-  private def shouldInlineDef(defn: OccurrenceAst.Def, ctx0: LocalContext): Boolean = {
-    !ctx0.currentlyInlining && !defn.context.isSelfRecursive && defn.context.isDirectCall
+  private def shouldInlineDef(defn: OccurrenceAst.Def, exps: List[Expr], ctx0: LocalContext): Boolean = {
+    !ctx0.currentlyInlining && !defn.context.isSelfRecursive &&
+      (defn.context.isDirectCall ||
+        (isHigherOrder(defn) && hasKnownLambda(defn.fparams, exps)))
+  }
+
+  /** Returns `true` if at least one formal parameter of `defn` has an arrow type. */
+  private def isHigherOrder(defn: OccurrenceAst.Def): Boolean = {
```

Comment:
I would argue that it is sufficient to check that one of the actual arguments is a lambda.